### PR TITLE
add placeholders

### PR DESCRIPTION
**Pull request by aquibyatoo**
### Status :
 In Development 

### Description :
Add placeholders for loading state for app. There are Box, Line, Image Card, Card and Paragraph placeholders.

##
### Related Issues :
#169

## Todos
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Review by rishichawda — CHANGES_REQUESTED** [review]
@aquibyatoo Can you add documentation so that it is possible for me to test it out quickly? You can refer other component directories to get an idea on how to document it.

File: lib/placeholders/box/readMe.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+## Navbar
+
+The box placeholder component.
+
+### Properties
+
+| Name     | Type     | Default   | Description                          |
+| :------- | :------- | :-------- | :----------------------------------- |
+| `square` | `circle` | &nbsp;    | Shapes of the box default is circle. |
```
Comment (rishichawda) [inline lib/placeholders/box/readMe.md]: I can't see `square` prop used anywhere in the component. Is this by mistake?

File: .gitignore [hunk 2]
```diff
@@ -6,4 +6,6 @@ coverage
 .nyc_output
 /dist
 /stats
-/components
\ No newline at end of file
+/components
+/.prettierrc.json
```
Comment (rishichawda) [inline .gitignore]: why was prettier used? We already have eslint set up and an eslintrc on the root.

**Comment by aquibyatoo** [comment]
yeah by mistake.  Its only shape there, fixed.

thanks

File: .gitignore (hunk 2 above)
Comment (rishichawda) [reply .gitignore]: Please make sure code is formatted using eslint and the config that is already present.

**Comment by aquibyatoo** [comment]
Actually its because i have configured prettier already and that is why i
put it in .gitignore.

thanks

File: lib/globals/_theme.scss [hunk 3]
```diff
@@ -55,3 +56,14 @@ $dark-hot-pink-20: rgba(225, 0, 80, 0.2);
 $orange-yellow-20: rgba(255, 165, 0, 0.2);
 $battleship-grey-20: rgba(108, 117, 125, 0.2);
 $dark-hot-pink-50: rgba(225, 0, 80, 0.5);
+
+$shimmer-color-light: #f6f7f8;
+$shimmer-color-dark: #edeef1;
+
+// *********** BOX SHADOWS ****************
```
Comment (rishichawda) [inline lib/globals/_theme.scss]: Comments for single lines can be provided at the end of the line. Comments like this increase the code length (Also try avoiding unnecessary line gaps). Moreover, The names can be descriptive to be understandable at first glance, so probably we can remove comments altogether and just use names that are descriptive enough.

File: lib/globals/animations.module.scss [hunk 4]
```diff
@@ -1,41 +1,75 @@
...
-    -webkit-animation-name: animateFocus;
-    animation-name: animateFocus;
-    @keyframes animateFocus {
-        0% {
-            width: 0%;
-        }
-        100% {
-            width: 100%;
-        }
+  -o-animation-name: animateFocus;
```
Comment (rishichawda) [inline lib/globals/animations.module.scss:2]: Looks like this was due to prettier. Can you try removing prettier and lint files with eslint and see if it reverts these changes?

File: lib/placeholders/box/readMe.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
+## Navbar
```
Comment (KRRISH96) [inline lib/placeholders/box/readMe.md]: I guess This is added while taking a markdown sample of other components.
Please replace with a relevant title for this component (ex: Box Placeholder).

File: .gitignore (hunk 2 above)
Comment (aquibyatoo) [reply .gitignore]: 321bd4ced83925829ae7a889f3d16964b6fa0c12

File: lib/globals/_theme.scss (hunk 3 above)
Comment (aquibyatoo) [reply lib/globals/_theme.scss]: 321bd4ced83925829ae7a889f3d16964b6fa0c12

File: lib/globals/animations.module.scss (hunk 4 above)
Comment (aquibyatoo) [reply lib/globals/animations.module.scss:2]: 321bd4ced83925829ae7a889f3d16964b6fa0c12

File: lib/placeholders/box/readMe.md (hunk 1 above)
Comment (aquibyatoo) [reply lib/placeholders/box/readMe.md]: 07e1bd894e713bbe71d96aaf554badda90d852db

File: lib/globals/animations.module.scss (hunk 4 above)
Comment (aquibyatoo) [reply lib/globals/animations.module.scss:2]: 0e0da4dd32e77d0afd2b322cf30853314dd207e8

File: lib/placeholders/box/readMe.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+const Demo = ({theme, shape}) => {
+    const classes = classnames(theme.styles, theme.boxStyles, theme[shape]);
+
+  const boxPlaceholderProps = {
+    className: classes,
+  };
+
+  return <div {...boxPlaceholderProps} />;
+}
```
Comment (KRRISH96) [inline lib/placeholders/box/readMe.md:73]: The Usage Section should be an example of how the component can be used.
```jsx
class Demo extends React.Component {
    state = {
       loading: false,
    }
    render() {
       const {loading} = this.state;
       return (
            <div>
                { loading ? <BoxPlaceholder /> : <div>User Content</div>
             </div>
        )
    }
}
```

Please have look at other components documentation to get an Idea.
Thanks.

File: lib/placeholders/card/index.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  theme: PropTypes.shape({}),
+  imageShape: PropTypes.oneOf(['default', 'circle']),
+};
+
+CardPlaceholder.defaultProps = {
+  theme: {},
+  imageShape: 'circle',
+};
+
+export default themr('CARPLACEHOLDER', defaultTheme)(CardPlaceholder);
```
Comment (KRRISH96) [inline lib/placeholders/card/index.js]: Looks to be a Typo, `CARDPLACEHOLDER`. 
This doesn't affect the component. Still good to avoid typos.

Comment (aquibyatoo) [reply lib/placeholders/card/index.js]: 977c4770419b2a5b4289c7e62a8812e8b16cf095

File: lib/placeholders/card/readMe.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
+## Navbar
```
Comment (KRRISH96) [inline lib/placeholders/card/readMe.md]: Please fix the title.

File: lib/placeholders/box/readMe.md (hunk 5 above)
Comment (aquibyatoo) [reply lib/placeholders/box/readMe.md]: 977c4770419b2a5b4289c7e62a8812e8b16cf095

File: lib/placeholders/card/readMe.md (hunk 8 above)
Comment (aquibyatoo) [reply lib/placeholders/card/readMe.md]: 977c4770419b2a5b4289c7e62a8812e8b16cf095

File: lib/placeholders/card/readMe.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        <BoxPlaceholder shape={imageShape} />
+      </div>
+
+      <LinePlaceholder lineWidth="100" />
+    </div>
+
+    <LinePlaceholder lineWidth="75" />
+    <LinePlaceholder lineWidth="50" />
+  </div>
+);
```
Comment (KRRISH96) [inline lib/placeholders/card/readMe.md]: As mentioned earlier, Please Update Usage here also.

You can have the above usage too, to show how the card placeholder can be constructed using `Box` and `LinePlaceholders` we have.


```jsx
class Demo extends React.Component {
    state = {
       loading: false,
    }
    render() {
       const {loading} = this.state;
       return (
            <div>
                { loading ? <CardPlaceholder /> : <div>User Card</div>
             </div>
        )
    }
}```

File: lib/placeholders/imageCard/readMe.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
+## Navbar
```
Comment (KRRISH96) [inline lib/placeholders/imageCard/readMe.md]: Please Fix title here and in all relevant MD files.

File: lib/placeholders/imageCard/readMe.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+  return (
+    <div {...ImageCardProps}>
+      <BoxPlaceholder shape={imageShape} />
+      {lines.map(line => (
+        <LinePlaceHolder lineWidth={line} />
+      ))}
+    </div>
+  );
+};
```
Comment (KRRISH96) [inline lib/placeholders/imageCard/readMe.md]: Similarly, Please Update usage section as mentioned earlier.
Thanks.

File: lib/placeholders/imageCard/readMe.md (hunk 10 above)
Comment (aquibyatoo) [reply lib/placeholders/imageCard/readMe.md]: 977c4770419b2a5b4289c7e62a8812e8b16cf095

**Review by KRRISH96 — CHANGES_REQUESTED** [review]
Please make changes as mentioned in Comments. Thanks.

File: lib/placeholders/paragraph/tests/accessibility.test.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+  it('should have only a single `div`', () => {
+    expect(actualValue('div')).equal(expectedValue);
+  });
+
+  it('should render the `Line Placeholder` along with props ', () => {
+    expect(wrappedComponent.contains(<div />)).to.equal(false);
+  });
+
+  it('should render the `Lard Placeholder` containing minimum 1 `div` element', () => {
```
Comment (KRRISH96) [inline lib/placeholders/paragraph/tests/accessibility.test.js]: Please Fix Typo, `Card Placeholder`.

File: lib/placeholders/line/tests/accessibility.test.js [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+  it('should have only a single `div`', () => {
+    expect(actualValue('div')).equal(expectedValue);
+  });
+
+  it('should render the `Line Placeholder` along with props ', () => {
+    expect(wrappedComponent.contains(<div />)).to.equal(false);
+  });
+
+  it('should render the `Lard Placeholder` containing minimum 1 `div` element', () => {
```
Comment (KRRISH96) [inline lib/placeholders/line/tests/accessibility.test.js]: Please fix Typo, `Card Placeholder`

File: lib/placeholders/box/readMe.md (hunk 6 above)
Comment (aquibyatoo) [reply lib/placeholders/box/readMe.md:73]: 1309bc5f877031f4e7719566dc665b2940b86281

File: lib/placeholders/card/readMe.md (hunk 9 above)
Comment (aquibyatoo) [reply lib/placeholders/card/readMe.md]: 1309bc5f877031f4e7719566dc665b2940b86281

File: lib/placeholders/imageCard/readMe.md (hunk 11 above)
Comment (aquibyatoo) [reply lib/placeholders/imageCard/readMe.md]: 1309bc5f877031f4e7719566dc665b2940b86281

File: lib/placeholders/paragraph/tests/accessibility.test.js (hunk 12 above)
Comment (aquibyatoo) [reply lib/placeholders/paragraph/tests/accessibility.test.js]: 1309bc5f877031f4e7719566dc665b2940b86281

File: lib/placeholders/line/tests/accessibility.test.js (hunk 13 above)
Comment (aquibyatoo) [reply lib/placeholders/line/tests/accessibility.test.js]: 1309bc5f877031f4e7719566dc665b2940b86281

**Review by rishichawda — CHANGES_REQUESTED** [review]
Regarding the functionalities and current configuration of the components : 

1. I can see there are two components with nearly same function : `ImageCard` and `CardPlaceholder`. I couldn't spot any functional difference between them and they look almost identical to each other too. 

2. Why do we have two exports : `ParagraphPlaceHolder` and `LinePlaceholder` - can it be done in a single component? 

3. `ParagraphPlaceHolder` seems to have box shadow. User might not want it in most of cases probably?

4. The "shimmer" effect looks like it is not completely carried out till the other end of the placeholders.

5. In `CardPlaceholder` the image looks like it has shimmer faster than the lines ( probably because lines are longer than the width of the card ). Can this be tweaked so it looks balanced?

6. The line widths in `ParagraphPlaceHolder` have a fixed value which increments as the lines proceed. Can this be randomised so that it looks more natural?

7. In `CardPlaceholder` component, the line widths are hard-coded. Can this be changed too? Can we also use the existing `Card` component instead of having a separate wrapper again? Since it has other options too like changing shadows, etc. Moreover, it would be then an easier content management from user perspective. Also, we won't have to ship a component like this altogether ( Can be put as a demo using `Card` and `LinePlaceholder` component.

**Review by rishichawda — COMMENTED** [review]
> Regarding the functionalities and current configuration of the components :
> 
> I can see there are two components with nearly same function : ImageCard and CardPlaceholder. I couldn't spot any functional difference between them and they look almost identical to each other too.
> 
> Why do we have two exports : ParagraphPlaceHolder and LinePlaceholder - can it be done in a single component?
> 
> ParagraphPlaceHolder seems to have box shadow. User might not want it in most of cases probably?
> 
> The "shimmer" effect looks like it is not completely carried out till the other end of the placeholders.
> 
> In CardPlaceholder the image looks like it has shimmer faster than the lines ( probably because lines are longer than the width of the card ). Can this be tweaked so it looks balanced?
> 
> The line widths in ParagraphPlaceHolder have a fixed value which increments as the lines proceed. Can this be randomised so that it looks more natural?
> 
> In CardPlaceholder component, the line widths are hard-coded. Can this be changed too? Can we also use the existing Card component instead of having a separate wrapper again? Since it has other options too like changing shadows, etc. Moreover, it would be then an easier content management from user perspective. Also, we won't have to ship a component like this altogether ( Can be put as a demo using Card and LinePlaceholder component.

@aquibyatoo Are all the changes present in this branch?

**Comment by aquibyatoo** [comment]
Closing this PR.

**Pull request closed without merging**